From 1f5f14d4f975bca88f65ba0e0dbd0a5f7ecdf1cf Mon Sep 17 00:00:00 2001
From: Ludovic Desroches <ludovic.desroches@atmel.com>
Date: Mon, 11 Feb 2013 17:53:11 +0100
Subject: [PATCH 309/319] ARM: at91: enhance linux-at91 at91sam9x5 support

This patch enhances the linux-at91 at91sam9x5 support mainly by updating
device tree files to use latest drivers and to enable dma.
at91_dt_defconfig has also been updated to support features available on
at91sam9x5 family.

Known issue:
- dma-rx for dbgu is not working
- no support for audio
- no support for ISI
- no support for dataflash memory on Cogent cpu modules

Signed-off-by: Ludovic Desroches <ludovic.desroches@atmel.com>
---
 arch/arm/boot/dts/at91sam9g15.dtsi  |    2 +-
 arch/arm/boot/dts/at91sam9g15ek.dts |   50 +++
 arch/arm/boot/dts/at91sam9g25ek.dts |   13 +-
 arch/arm/boot/dts/at91sam9g35.dtsi  |    2 +-
 arch/arm/boot/dts/at91sam9g35ek.dts |   32 +-
 arch/arm/boot/dts/at91sam9x25ek.dts |   14 +-
 arch/arm/boot/dts/at91sam9x35.dtsi  |    2 +-
 arch/arm/boot/dts/at91sam9x35ek.dts |   58 +++
 arch/arm/boot/dts/at91sam9x5.dtsi   |  750 +++++++++++++++++++++--------------
 arch/arm/boot/dts/at91sam9x5cm.dtsi |   15 +
 arch/arm/boot/dts/at91sam9x5ek.dtsi |   21 +-
 arch/arm/configs/at91_dt_defconfig  |   33 +-
 arch/arm/mach-at91/at91sam9x5.c     |   33 +-
 drivers/net/can/Kconfig             |    2 +-
 14 files changed, 692 insertions(+), 335 deletions(-)

diff --git a/arch/arm/boot/dts/at91sam9g15.dtsi b/arch/arm/boot/dts/at91sam9g15.dtsi
index fbe7a70..de28baa 100644
--- a/arch/arm/boot/dts/at91sam9g15.dtsi
+++ b/arch/arm/boot/dts/at91sam9g15.dtsi
@@ -18,7 +18,7 @@
 				atmel,mux-mask = <
 				      /*    A         B          C     */
 				       0xffffffff 0xffe0399f 0x00000000  /* pioA */
-				       0x00040000 0x00047e3f 0x00000000  /* pioB */
+				       0x0007ffc0 0x00047e3f 0x00000000  /* pioB */
 				       0xfdffffff 0x00000000 0xb83fffff  /* pioC */
 				       0x003fffff 0x003f8000 0x00000000  /* pioD */
 				      >;
diff --git a/arch/arm/boot/dts/at91sam9g15ek.dts b/arch/arm/boot/dts/at91sam9g15ek.dts
index 86dd3f6..756c931 100644
--- a/arch/arm/boot/dts/at91sam9g15ek.dts
+++ b/arch/arm/boot/dts/at91sam9g15ek.dts
@@ -13,4 +13,54 @@
 / {
 	model = "Atmel AT91SAM9G25-EK";
 	compatible = "atmel,at91sam9g15ek", "atmel,at91sam9x5ek", "atmel,at91sam9x5", "atmel,at91sam9";
+
+	ahb {
+		apb {
+			spi0: spi@f0000000 {
+				status = "okay";
+			};
+
+			i2c0: i2c@f8010000 {
+				status = "okay";
+
+				qt1070: keyboard@1b {
+					compatible = "qt1070";
+					reg = <0x1b>;
+					interrupt-parent = <&pioA>;
+					interrupts = <7 0x0>;
+					pinctrl-names = "default";
+					pinctrl-0 = <&pinctrl_qt1070_irq>;
+				};
+			};
+
+			lcd_bus@f8038000 {
+				status = "okay";
+
+				lcd@f8038000 {
+					status = "okay";
+				};
+
+				lcdovl1@f8038100 {
+					status = "okay";
+				};
+
+				lcdheo1@f8038280 {
+					status = "okay";
+				};
+			};
+
+			tsadcc: tsadcc@f804c000 {
+				status = "okay";
+			};
+
+			pinctrl@fffff400 {
+				board {
+					pinctrl_qt1070_irq: qt1070_irq {
+						atmel,pins =
+							<0 7 0x0 0x5>; /* PA7 GPIO with pull up deglith */
+					};
+				};
+			};
+		};
+	};
 };
diff --git a/arch/arm/boot/dts/at91sam9g25ek.dts b/arch/arm/boot/dts/at91sam9g25ek.dts
index 0c917e8..5ae11a3 100644
--- a/arch/arm/boot/dts/at91sam9g25ek.dts
+++ b/arch/arm/boot/dts/at91sam9g25ek.dts
@@ -16,17 +16,16 @@
 
 	ahb {
 		apb {
-			mmc1: mmc@f000c000 {
+			spi0: spi@f0000000 {
 				status = "okay";
 			};
 
-			spi0: spi@f0000000 {
+			i2c0: i2c@f8010000 {
+				status = "okay";
+			};
+
+			macb0: ethernet@f802c000 {
 				status = "okay";
-				m25p80@0 {
-					compatible = "atmel,at25df321a";
-					spi-max-frequency = <50000000>;
-					reg = <0>;
-				};
 			};
 
 			watchdog@fffffe40 {
diff --git a/arch/arm/boot/dts/at91sam9g35.dtsi b/arch/arm/boot/dts/at91sam9g35.dtsi
index f9d14a7..d7b0ea5 100644
--- a/arch/arm/boot/dts/at91sam9g35.dtsi
+++ b/arch/arm/boot/dts/at91sam9g35.dtsi
@@ -18,7 +18,7 @@
 				atmel,mux-mask = <
 				      /*    A         B          C     */
 				       0xffffffff 0xffe0399f 0xc000000c  /* pioA */
-				       0x000406ff 0x00047e3f 0x00000000  /* pioB */
+				       0x0007ffff 0x00047e3f 0x00000000  /* pioB */
 				       0xfdffffff 0x00000000 0xb83fffff  /* pioC */
 				       0x003fffff 0x003f8000 0x00000000  /* pioD */
 				      >;
diff --git a/arch/arm/boot/dts/at91sam9g35ek.dts b/arch/arm/boot/dts/at91sam9g35ek.dts
index 9f91dbc..5eec1edf 100644
--- a/arch/arm/boot/dts/at91sam9g35ek.dts
+++ b/arch/arm/boot/dts/at91sam9g35ek.dts
@@ -16,7 +16,24 @@
 
 	ahb {
 		apb {
-			mmc1: mmc@f000c000 {
+			spi0: spi@f0000000 {
+				status = "okay";
+			};
+
+			i2c0: i2c@f8010000 {
+				status = "okay";
+
+				qt1070: keyboard@1b {
+					compatible = "qt1070";
+					reg = <0x1b>;
+					interrupt-parent = <&pioA>;
+					interrupts = <7 0x0>;
+					pinctrl-names = "default";
+					pinctrl-0 = <&pinctrl_qt1070_irq>;
+				};
+			};
+
+			macb0: ethernet@f802c000 {
 				status = "okay";
 			};
 
@@ -34,6 +51,19 @@
 					status = "okay";
 				};
 			};
+
+			tsadcc: tsadcc@f804c000 {
+				status = "okay";
+			};
+
+			pinctrl@fffff400 {
+				board {
+					pinctrl_qt1070_irq: qt1070_irq {
+						atmel,pins =
+							<0 7 0x0 0x5>; /* PA7 GPIO with pull up deglith */
+					};
+				};
+			};
 		};
 	};
 };
diff --git a/arch/arm/boot/dts/at91sam9x25ek.dts b/arch/arm/boot/dts/at91sam9x25ek.dts
index 7e38951..64e3e8c 100644
--- a/arch/arm/boot/dts/at91sam9x25ek.dts
+++ b/arch/arm/boot/dts/at91sam9x25ek.dts
@@ -16,7 +16,19 @@
 
 	ahb {
 		apb {
-			mmc1: mmc@f000c000 {
+			spi0: spi@f0000000 {
+				status = "okay";
+			};
+
+			can1: can@f8004000 {
+				status = "okay";
+			};
+
+			i2c0: i2c@f8010000 {
+				status = "okay";
+			};
+
+			macb0: ethernet@f802c000 {
 				status = "okay";
 			};
 
diff --git a/arch/arm/boot/dts/at91sam9x35.dtsi b/arch/arm/boot/dts/at91sam9x35.dtsi
index fb102d6..25bafb5 100644
--- a/arch/arm/boot/dts/at91sam9x35.dtsi
+++ b/arch/arm/boot/dts/at91sam9x35.dtsi
@@ -18,7 +18,7 @@
 				atmel,mux-mask = <
 				      /*    A         B          C     */
 				       0xffffffff 0xffe03fff 0xc000000c  /* pioA */
-				       0x000406ff 0x00047e3f 0x00000000  /* pioB */
+				       0x0007ffff 0x00047e3f 0x00000000  /* pioB */
 				       0xfdffffff 0x00000000 0xb83fffff  /* pioC */
 				       0x003fffff 0x003f8000 0x00000000  /* pioD */
 				      >;
diff --git a/arch/arm/boot/dts/at91sam9x35ek.dts b/arch/arm/boot/dts/at91sam9x35ek.dts
index 5ccb607..35a3f8b 100644
--- a/arch/arm/boot/dts/at91sam9x35ek.dts
+++ b/arch/arm/boot/dts/at91sam9x35ek.dts
@@ -13,4 +13,62 @@
 / {
 	model = "Atmel AT91SAM9X35-EK";
 	compatible = "atmel,at91sam9x35ek", "atmel,at91sam9x5ek", "atmel,at91sam9x5", "atmel,at91sam9";
+
+	ahb {
+		apb {
+			spi0: spi@f0000000 {
+				status = "okay";
+			};
+
+			can1: can@f8004000 {
+				status = "okay";
+			};
+
+			i2c0: i2c@f8010000 {
+				status = "okay";
+
+				qt1070: keyboard@1b {
+					compatible = "qt1070";
+					reg = <0x1b>;
+					interrupt-parent = <&pioA>;
+					interrupts = <7 0x0>;
+					pinctrl-names = "default";
+					pinctrl-0 = <&pinctrl_qt1070_irq>;
+				};
+			};
+
+			macb0: ethernet@f802c000 {
+				status = "okay";
+			};
+
+			lcd_bus@f8038000 {
+				status = "okay";
+
+				lcd@f8038000 {
+					status = "okay";
+				};
+
+				lcdovl1@f8038100 {
+					status = "okay";
+				};
+
+				lcdheo1@f8038280 {
+					status = "okay";
+				};
+			};
+
+			tsadcc: tsadcc@f804c000 {
+				status = "okay";
+			};
+
+			pinctrl@fffff400 {
+				board {
+					pinctrl_qt1070_irq: qt1070_irq {
+						atmel,pins =
+							<0 7 0x0 0x5>; /* PA7 GPIO with pull up deglith */
+					};
+				};
+			};
+		};
+	};
 };
diff --git a/arch/arm/boot/dts/at91sam9x5.dtsi b/arch/arm/boot/dts/at91sam9x5.dtsi
index 756c9c1..11412c2 100644
--- a/arch/arm/boot/dts/at91sam9x5.dtsi
+++ b/arch/arm/boot/dts/at91sam9x5.dtsi
@@ -53,38 +53,80 @@
 			#size-cells = <1>;
 			ranges;
 
-			aic: interrupt-controller@fffff000 {
-				#interrupt-cells = <3>;
-				compatible = "atmel,at91rm9200-aic";
-				interrupt-controller;
-				reg = <0xfffff000 0x200>;
-				atmel,external-irqs = <31>;
+			spi0: spi@f0000000 {
+				#address-cells = <1>;
+				#size-cells = <0>;
+				compatible = "atmel,at91sam9x5-spi";
+				reg = <0xf0000000 0x100>;
+				interrupts = <13 4 3>;
+				pinctrl-names = "default";
+				pinctrl-0 = <&pinctrl_spi0>;
+				dma-mask = <0xffffffff>;
+				dma = <&dma0 0x10002212>;
+				status = "disabled";
 			};
 
-			ramc0: ramc@ffffe800 {
-				compatible = "atmel,at91sam9g45-ddramc";
-				reg = <0xffffe800 0x200>;
+			spi1: spi@f0004000 {
+				#address-cells = <1>;
+				#size-cells = <0>;
+				compatible = "atmel,at91sam9x5-spi";
+				reg = <0xf0004000 0x100>;
+				interrupts = <14 4 3>;
+				pinctrl-names = "default";
+				pinctrl-0 = <&pinctrl_spi1>;
+				dma-mask = <0xffffffff>;
+				dma = <&dma1 0x10002212>;
+				status = "disabled";
 			};
 
-			pmc: pmc@fffffc00 {
-				compatible = "atmel,at91rm9200-pmc";
-				reg = <0xfffffc00 0x100>;
+			mmc0: mmc@f0008000 {
+				compatible = "atmel,hsmci";
+				reg = <0xf0008000 0x600>;
+				interrupts = <12 4 0>;
+				status = "disabled";
+				#address-cells = <1>;
+				#size-cells = <0>;
+				dma = <&dma0 0x10002200>;
+				pinctrl-names = "default";
 			};
 
-			rstc@fffffe00 {
-				compatible = "atmel,at91sam9g45-rstc";
-				reg = <0xfffffe00 0x10>;
+			mmc1: mmc@f000c000 {
+				compatible = "atmel,hsmci";
+				reg = <0xf000c000 0x600>;
+				interrupts = <26 4 0>;
+				status = "disabled";
+				#address-cells = <1>;
+				#size-cells = <0>;
+				dma = <&dma1 0x10002200>;
+				pinctrl-names = "default";
 			};
 
-			shdwc@fffffe10 {
-				compatible = "atmel,at91sam9x5-shdwc";
-				reg = <0xfffffe10 0x10>;
+			ssc0: ssc@f0010000 {
+				compatible = "atmel,at91sam9g45-ssc";
+				reg = <0xf0010000 0x4000>;
+				interrupts = <28 4 4>;
+				pinctrl-names = "default";
+				pinctrl-0 = <&pinctrl_ssc0_tx &pinctrl_ssc0_rx>;
+				dma = <&dma0 0x100022de>;
+				status = "disabled";
 			};
 
-			pit: timer@fffffe30 {
-				compatible = "atmel,at91sam9260-pit";
-				reg = <0xfffffe30 0xf>;
-				interrupts = <1 4 7>;
+			can0: can@f8000000 {
+				compatible = "atmel,at91sam9x5-can";
+				reg = <0xf8000000 0x300>;
+				interrupts = <29 4 3>;
+				pinctrl-names = "default";
+				pinctrl-0 = <&pinctrl_can0_rx_tx>;
+				status = "disabled";
+			};
+
+			can1: can@f8004000 {
+				compatible = "atmel,at91sam9x5-can";
+				reg = <0xf8004000 0x300>;
+				interrupts = <30 4 3>;
+				pinctrl-names = "default";
+				pinctrl-0 = <&pinctrl_can1_rx_tx>;
+				status = "disabled";
 			};
 
 			tcb0: timer@f8008000 {
@@ -99,6 +141,188 @@
 				interrupts = <17 4 0>;
 			};
 
+			i2c0: i2c@f8010000 {
+				compatible = "atmel,at91sam9x5-i2c";
+				reg = <0xf8010000 0x100>;
+				interrupts = <9 4 6>;
+				dma = <&dma0 0x10002278>;
+				#address-cells = <1>;
+				#size-cells = <0>;
+				pinctrl-names = "default";
+				pinctrl-0 = <&pinctrl_i2c0>;
+				status = "disabled";
+			};
+
+			i2c1: i2c@f8014000 {
+				compatible = "atmel,at91sam9x5-i2c";
+				reg = <0xf8014000 0x100>;
+				interrupts = <10 4 6>;
+				dma = <&dma1 0x10002256>;
+				#address-cells = <1>;
+				#size-cells = <0>;
+				pinctrl-names = "default";
+				pinctrl-0 = <&pinctrl_i2c1>;
+				status = "disabled";
+			};
+
+			i2c2: i2c@f8018000 {
+				compatible = "atmel,at91sam9x5-i2c";
+				reg = <0xf8018000 0x100>;
+				interrupts = <11 4 6>;
+				dma = <&dma0 0x1000229A>;
+				#address-cells = <1>;
+				#size-cells = <0>;
+				pinctrl-names = "default";
+				pinctrl-0 = <&pinctrl_i2c2>;
+				status = "disabled";
+			};
+
+			usart0: serial@f801c000 {
+				compatible = "atmel,at91sam9x5-usart";
+				reg = <0xf801c000 0x200>;
+				interrupts = <5 4 5>;
+				atmel,use-dma-rx;
+				dma-rx = <&dma0 0x20000204>;
+				atmel,use-dma-tx;
+				dma-tx = <&dma0 0x10002030>;
+				pinctrl-names = "default";
+				pinctrl-0 = <&pinctrl_usart0>;
+				status = "disabled";
+			};
+
+			usart1: serial@f8020000 {
+				compatible = "atmel,at91sam9x5-usart";
+				reg = <0xf8020000 0x200>;
+				interrupts = <6 4 5>;
+				atmel,use-dma-rx;
+				dma-rx = <&dma0 0x20000206>;
+				atmel,use-dma-tx;
+				dma-tx = <&dma0 0x10002050>;
+				pinctrl-names = "default";
+				pinctrl-0 = <&pinctrl_usart1>;
+				status = "disabled";
+			};
+
+			usart2: serial@f8024000 {
+				compatible = "atmel,at91sam9x5-usart";
+				reg = <0xf8024000 0x200>;
+				interrupts = <7 4 5>;
+				atmel,use-dma-rx;
+				dma-rx = <&dma1 0x2000020d>;
+				atmel,use-dma-tx;
+				dma-tx = <&dma1 0x100020c0>;
+				pinctrl-names = "default";
+				pinctrl-0 = <&pinctrl_usart2>;
+				status = "disabled";
+			};
+
+			macb0: ethernet@f802c000 {
+				compatible = "cdns,at32ap7000-macb", "cdns,macb";
+				reg = <0xf802c000 0x100>;
+				interrupts = <24 4 3>;
+				pinctrl-names = "default";
+				pinctrl-0 = <&pinctrl_macb0_rmii>;
+				status = "disabled";
+			};
+
+			macb1: ethernet@f8030000 {
+				compatible = "cdns,at32ap7000-macb", "cdns,macb";
+				reg = <0xf8030000 0x100>;
+				interrupts = <27 4 3>;
+				pinctrl-names = "default";
+				pinctrl-0 = <&pinctrl_macb1_rmii>;
+				status = "disabled";
+			};
+
+			lcd_bus@f8038000 {
+				#address-cells = <1>;
+				#size-cells = <1>;
+				compatible = "atmel,at91sam9x5-lcd-bus", "simple-bus";
+				ranges = <0xf8038000 0xf8038000 0x4000>;
+				pinctrl-names = "default";
+				pinctrl-0 = <&pinctrl_lcd>;
+				status = "disabled";
+
+				lcd@f8038000 {
+					compatible = "atmel,at91sam9x5-lcd";
+					reg = <0xf8038000 0xff
+					       0xf8038400 0x3ff>;
+					interrupts = <25 4 3>;
+					status = "disabled";
+				};
+
+				lcdovl1@f8038100 {
+					compatible = "atmel,at91sam9x5-lcd";
+					reg = <0xf8038100 0xff
+					       0xf8038800 0x3ff>;
+					status = "disabled";
+				};
+
+				lcdheo1@f8038280 {
+					compatible = "atmel,at91sam9x5-heo";
+					reg = <0xf8038280 0xbf>;
+					interrupts = <25 4 3>;
+					status = "disabled";
+				};
+			};
+
+			tsadcc: tsadcc@f804c000 {
+				compatible = "atmel,at91sam9x5-tsadcc";
+				reg = <0xf804c000 0x4000>;
+				interrupts = <19 4 5>;
+				atmel,tsadcc_clock = <300000>;
+				atmel,filtering_average = <0x03>;
+				atmel,pendet_debounce = <0x08>;
+				atmel,pendet_sensitivity = <0x02>;
+				atmel,ts_sample_hold_time = <0x0a>;
+				status = "disabled";
+			};
+
+			adc0: adc@f804c000 {
+				compatible = "atmel,at91sam9260-adc";
+				reg = <0xf804c000 0x100>;
+				interrupts = <19 4 0>;
+				pinctrl-names = "default";
+				atmel,adc-use-external;
+				atmel,adc-num-channels = <12>;
+				atmel,adc-startup-time = <40>;
+				atmel,adc-channel-base = <0x50>;
+				atmel,adc-drdy-mask = <0x1000000>;
+				atmel,adc-status-register = <0x30>;
+				atmel,adc-trigger-register = <0xc0>;
+				atmel,adc-res = <8 10>;
+				atmel,adc-res-names = "lowres", "highres";
+				status = "disabled";
+
+				trigger@0 {
+					trigger-name = "external-rising";
+					trigger-value = <0x1>;
+					trigger-external;
+				};
+
+				trigger@1 {
+					trigger-name = "external-falling";
+					trigger-value = <0x2>;
+					trigger-external;
+				};
+
+				trigger@2 {
+					trigger-name = "external-any";
+					trigger-value = <0x3>;
+					trigger-external;
+				};
+
+				trigger@3 {
+					trigger-name = "continuous";
+					trigger-value = <0x6>;
+				};
+			};
+
+			ramc0: ramc@ffffe800 {
+				compatible = "atmel,at91sam9g45-ddramc";
+				reg = <0xffffe800 0x200>;
+			};
+
 			dma0: dma-controller@ffffec00 {
 				compatible = "atmel,at91sam9g45-dma";
 				reg = <0xffffec00 0x200>;
@@ -113,13 +337,109 @@
 				#dma-cells = <1>;
 			};
 
-			pinctrl@fffff400 {
-				#address-cells = <1>;
-				#size-cells = <1>;
-				compatible = "atmel,at91sam9x5-pinctrl", "atmel,at91rm9200-pinctrl", "simple-bus";
-				ranges = <0xfffff400 0xfffff400 0x800>;
+			aic: interrupt-controller@fffff000 {
+				#interrupt-cells = <3>;
+				compatible = "atmel,at91rm9200-aic";
+				interrupt-controller;
+				reg = <0xfffff000 0x200>;
+				atmel,external-irqs = <31>;
+			};
+
+			dbgu: serial@fffff200 {
+				compatible = "atmel,at91sam9x5-usart";
+				reg = <0xfffff200 0x200>;
+				interrupts = <1 4 7>;
+				/*atmel,use-dma-rx;
+				dma-rx = <&dma1 0x20000209>;*/
+				atmel,use-dma-tx;
+				dma-tx = <&dma1 0x10002080>;
+				pinctrl-names = "default";
+				pinctrl-0 = <&pinctrl_dbgu>;
+				status = "disabled";
+			};
+
+			pinctrl@fffff400 {
+				#address-cells = <1>;
+				#size-cells = <1>;
+				compatible = "atmel,at91sam9x5-pinctrl", "atmel,at91rm9200-pinctrl", "simple-bus";
+				ranges = <0xfffff400 0xfffff400 0x800>;
+
+				/* shared pinctrl settings */
+				adc0 {
+					pinctrl_adc0_adtrg: adc0_adtrg {
+						atmel,pins =
+							<1 18 0x2 0x0>;	/* PB18 periph B ADTRG */
+					};
+					/*
+					 * Hardware don't care about the following pin configurations
+					 * but it has to be done to prevent pin request from other drivers.
+					 */
+					pinctrl_adc0_ad0: adc0_ad0 {
+						atmel,pins =
+							<1 11 0x1 0x0>; /* PB11 alternate AD0 */
+					};
+					pinctrl_adc0_ad1: adc0_ad1 {
+						atmel,pins =
+							<1 12 0x1 0x0>; /* PB12 alternate AD1 */
+					};
+					pinctrl_adc0_ad2: adc0_ad2 {
+						atmel,pins =
+							<1 13 0x1 0x0>; /* PB13 alternate AD2 */
+					};
+					pinctrl_adc0_ad3: adc0_ad3 {
+						atmel,pins =
+							<1 14 0x1 0x0>; /* PB14 alternate AD3 */
+					};
+					pinctrl_adc0_ad4: adc0_ad4 {
+						atmel,pins =
+							<1 15 0x1 0x0>; /* PB15 alternate AD4 */
+					};
+					pinctrl_adc0_ad5: adc0_ad5 {
+						atmel,pins =
+							<1 16 0x1 0x0>; /* PB16 alternate AD5 */
+					};
+					pinctrl_adc0_ad6: adc0_ad6 {
+						atmel,pins =
+							<1 17 0x1 0x0>; /* PB17 alternate AD6 */
+					};
+					pinctrl_adc0_ad7: adc0_ad7 {
+						atmel,pins =
+							<1 6 0x1 0x0>; /* PB6 alternate AD7 */
+					};
+					pinctrl_adc0_ad8: adc0_ad8 {
+						atmel,pins =
+							<1 7 0x1 0x0>; /* PB7 alternate AD8 */
+					};
+					pinctrl_adc0_ad9: adc0_ad9 {
+						atmel,pins =
+							<1 8 0x1 0x0>; /* PB8 alternate AD9 */
+					};
+					pinctrl_adc0_ad10: adc0_ad10 {
+						atmel,pins =
+							<1 9 0x1 0x0>; /* PB9 alternate AD10 */
+					};
+					pinctrl_adc0_ad11: adc0_ad11 {
+						atmel,pins =
+							<1 10 0x1 0x0>; /* PB10 alternate AD11 */
+					};
+				};
+
+				can0 {
+					pinctrl_can0_rx_tx: can0_rx_tx {
+						atmel,pins =
+							<0 9 0x2 0x0	/* PA9 periph B CANRX0 */
+							 0 10 0x2 0x0>;	/* PA10 periph B CANTX0 */
+					};
+				};
+
+				can1 {
+					pinctrl_can1_rx_tx: can1_rx_tx {
+						atmel,pins =
+							<0 5 0x2 0x0	/* PA5 periph B CANTX1 */
+							 0 6 0x2 0x0>;	/* PA6 periph B CANRX1 */
+					};
+				};
 
-				/* shared pinctrl settings */
 				dbgu {
 					pinctrl_dbgu: dbgu-0 {
 						atmel,pins =
@@ -128,72 +448,140 @@
 					};
 				};
 
-				uart0 {
-					pinctrl_uart0: uart0-0 {
+				macb0 {
+					pinctrl_macb0_rmii: macb0_rmii-0 {
+						atmel,pins =
+							<1 0 0x1 0x0	/* PB0 periph A ERX0 */
+							 1 1 0x1 0x0	/* PB1 periph A ERX1 */
+							 1 2 0x1 0x0	/* PB2 periph A ERXER */
+							 1 3 0x1 0x0	/* PB3 periph A ERXDV */
+							 1 4 0x1 0x0	/* PB4 periph A ETXCK */
+							 1 5 0x1 0x0	/* PB5 periph A EMDIO */
+							 1 6 0x1 0x0	/* PB6 periph A EMDC */
+							 1 7 0x1 0x0	/* PB7 periph A ETXEN */
+							 1 9 0x1 0x0	/* PB9 periph A ETX0 */
+							 1 10 0x1 0x0>;	/* PB10 periph A ETX1 */
+					};
+				};
+
+				macb1 {
+					pinctrl_macb1_rmii: macb1_rmii-0 {
+						atmel,pins =
+							<2 20 0x2 0x0	/* PC20 periph B E1_RX0 */
+							 2 21 0x2 0x0	/* PC21 periph B E1_RX1 */
+							 2 16 0x2 0x0	/* PC16 periph B E1_RXER */
+							 2 28 0x2 0x0	/* PC28 periph B E1_CRSDV */
+							 2 29 0x2 0x0	/* PC29 periph B E1_TXCK */
+							 2 31 0x2 0x0	/* PC31 periph B E1_MDIO */
+							 2 30 0x2 0x0	/* PC30 periph B E1_MDC */
+							 2 27 0x2 0x0	/* PC27 periph B E1_TXEN */
+							 2 18 0x2 0x0	/* PC18 periph B E1_TX0 */
+							 2 19 0x2 0x0>;	/* PC19 periph B E1_TX1 */
+					};
+				};
+
+				spi0 {
+					pinctrl_spi0: spi0-0 {
+						atmel,pins =
+							<0 11 0x1 0x0	/* PA11 periph A SPI0_MISO */
+							 0 12 0x1 0x0	/* PA12 periph A SPI0_MOSI */
+							 0 13 0x1 0x0	/* PA13 periph A SPI0_SPCK */
+							 0 14 0x1 0x0>;	/* PA14 periph A SPI0_NPCS0 */
+					};
+				};
+
+				spi1 {
+					pinctrl_spi1: spi1-0 {
+						atmel,pins =
+							<0 21 0x2 0x0	/* PA21 periph B SPI1_MISO */
+							 0 22 0x2 0x0	/* PA12 periph B SPI1_MOSI */
+							 0 23 0x2 0x0	/* PA23 periph B SPI1_SPCK */
+							 0 8 0x2 0x0>;	/* PA8 periph B SPI1_NPCS0 */
+					};
+				};
+
+				ssc0 {
+					pinctrl_ssc0_tx: ssc0_tx {
+						atmel,pins =
+							<0 24 0x2 0x0	/* PA24 periph B TK */
+							 0 25 0x2 0x0	/* PA25 periph B TF */
+							 0 26 0x2 0x0>;	/* PA26 periph B TD */
+					};
+
+					pinctrl_ssc0_rx: ssc0_rx {
+						atmel,pins =
+							<0 27 0x2 0x0	/* PA27 periph B RD */
+							 0 28 0x2 0x0	/* PA28 periph B RK */
+							 0 29 0x2 0x0>;	/* PA29 periph B RF */
+					};
+				};
+
+				usart0 {
+					pinctrl_usart0: usart0-0 {
 						atmel,pins =
 							<0 0 0x1 0x1	/* PA0 periph A with pullup */
 							 0 1 0x1 0x0>;	/* PA1 periph A */
 					};
 
-					pinctrl_uart0_rts_cts: uart0_rts_cts-0 {
+					pinctrl_usart0_rts_cts: usart0_rts_cts-0 {
 						atmel,pins =
 							<0 2 0x1 0x0	/* PA2 periph A */
 							 0 3 0x1 0x0>;	/* PA3 periph A */
 					};
 				};
 
-				uart1 {
-					pinctrl_uart1: uart1-0 {
+				usart1 {
+					pinctrl_usart1: usart1-0 {
 						atmel,pins =
 							<0 5 0x1 0x1	/* PA5 periph A with pullup */
 							 0 6 0x1 0x0>;	/* PA6 periph A */
 					};
 
-					pinctrl_uart1_rts_cts: uart1_rts_cts-0 {
+					pinctrl_usart1_rts_cts: usart1_rts_cts-0 {
 						atmel,pins =
 							<2 27 0x3 0x0	/* PC27 periph C */
 							 2 28 0x3 0x0>;	/* PC28 periph C */
 					};
 				};
 
-				uart2 {
-					pinctrl_uart2: uart2-0 {
+				usart2 {
+					pinctrl_usart2: usart2-0 {
 						atmel,pins =
 							<0 7 0x1 0x1	/* PA7 periph A with pullup */
 							 0 8 0x1 0x0>;	/* PA8 periph A */
 					};
 
-					pinctrl_uart2_rts_cts: uart2_rts_cts-0 {
+					pinctrl_usart2_rts_cts: usart2_rts_cts-0 {
 						atmel,pins =
 							<1 0 0x2 0x0	/* PB0 periph B */
 							 1 1 0x2 0x0>;	/* PB1 periph B */
 					};
 				};
 
-				uart3 {
-					pinctrl_uart3: uart3-0 {
+				usart3 {
+					pinctrl_usart3: usart3-0 {
 						atmel,pins =
 							<2 22 0x2 0x1	/* PC22 periph B with pullup */
 							 2 23 0x2 0x0>;	/* PC23 periph B */
 					};
 
-					pinctrl_uart3_rts_cts: uart3_rts_cts-0 {
+					pinctrl_usart3_rts_cts: usart3_rts_cts-0 {
 						atmel,pins =
 							<2 24 0x2 0x0	/* PC24 periph B */
 							 2 25 0x2 0x0>;	/* PC25 periph B */
 					};
 				};
 
-				usart0 {
-					pinctrl_usart0: usart0-0 {
+				uart0 {
+					pinctrl_uart0: uart0-0 {
 						atmel,pins =
 							<2 8 0x3 0x0	/* PC8 periph C */
 							 2 9 0x3 0x1>;	/* PC9 periph C with pullup */
 					};
 				};
 
-				usart1 {
-					pinctrl_usart1: usart1-0 {
+				uart1 {
+					pinctrl_uart1: uart1-0 {
 						atmel,pins =
 							<2 16 0x3 0x0	/* PC16 periph C */
 							 2 17 0x3 0x1>;	/* PC17 periph C with pullup */
@@ -243,24 +631,24 @@
 				i2c0 {
 					pinctrl_i2c0: i2c0-0 {
 						atmel,pins =
-							<0 30 0x1 0x2	/* PA30 periph A Multidrive */
-							 0 31 0x1 0x3>;	/* PA31 periph A Multidrive with pullup */
+							<0 30 0x1 0x0	/* PA30 periph A TWD0 */
+							 0 31 0x1 0x0>;	/* PA31 periph A TWCK0 */
 					};
 				};
 
 				i2c1 {
 					pinctrl_i2c1: i2c1-0 {
 						atmel,pins =
-							<2 0 0x3 0x2	/* PC0 periph C Multidrive */
-							 2 1 0x3 0x3>;	/* PC1 periph C Multidrive with pullup */
+							<2 0 0x3 0x0	/* PC0 periph C TWD1 */
+							 2 1 0x3 0x0>;	/* PC1 periph C TWCK1 */
 					};
 				};
 
 				i2c2 {
 					pinctrl_i2c2: i2c2-0 {
 						atmel,pins =
-							<1 4 0x2 0x2	/* PB4 periph B Multidrive */
-							 1 5 0x2 0x3>;	/* PB5 periph B Multidrive with pullup */
+							<1 4 0x2 0x0	/* PB4 periph B TWD2 */
+							 1 5 0x2 0x0>;	/* PB5 periph B TWCK2 */
 					};
 				};
 
@@ -375,222 +763,25 @@
 				};
 			};
 
-			dbgu: serial@fffff200 {
-				compatible = "atmel,at91sam9260-usart";
-				reg = <0xfffff200 0x200>;
-				interrupts = <1 4 7>;
-				pinctrl-names = "default";
-				pinctrl-0 = <&pinctrl_dbgu>;
-				status = "disabled";
-			};
-
-			usart0: serial@f801c000 {
-				compatible = "atmel,at91sam9260-usart";
-				reg = <0xf801c000 0x200>;
-				interrupts = <5 4 5>;
-				atmel,use-dma-rx;
-				atmel,use-dma-tx;
-				pinctrl-names = "default";
-				pinctrl-0 = <&pinctrl_uart0>;
-				status = "disabled";
-			};
-
-			usart1: serial@f8020000 {
-				compatible = "atmel,at91sam9260-usart";
-				reg = <0xf8020000 0x200>;
-				interrupts = <6 4 5>;
-				atmel,use-dma-rx;
-				atmel,use-dma-tx;
-				pinctrl-names = "default";
-				pinctrl-0 = <&pinctrl_uart1>;
-				status = "disabled";
-			};
-
-			usart2: serial@f8024000 {
-				compatible = "atmel,at91sam9260-usart";
-				reg = <0xf8024000 0x200>;
-				interrupts = <7 4 5>;
-				atmel,use-dma-rx;
-				atmel,use-dma-tx;
-				pinctrl-names = "default";
-				pinctrl-0 = <&pinctrl_uart2>;
-				status = "disabled";
-			};
-
-			macb0: ethernet@f802c000 {
-				compatible = "cdns,at32ap7000-macb", "cdns,macb";
-				reg = <0xf802c000 0x100>;
-				interrupts = <24 4 3>;
-				status = "disabled";
-			};
-
-			macb1: ethernet@f8030000 {
-				compatible = "cdns,at32ap7000-macb", "cdns,macb";
-				reg = <0xf8030000 0x100>;
-				interrupts = <27 4 3>;
-				status = "disabled";
-			};
-
-			i2c0: i2c@f8010000 {
-				compatible = "atmel,at91sam9x5-i2c";
-				reg = <0xf8010000 0x100>;
-				interrupts = <9 4 6>;
-				dma = <&dma0 0x10002278>;
-				#address-cells = <1>;
-				#size-cells = <0>;
-				pinctrl-names = "default";
-				pinctrl-0 = <&pinctrl_i2c0>;
-				status = "disabled";
-			};
-
-			i2c1: i2c@f8014000 {
-				compatible = "atmel,at91sam9x5-i2c";
-				reg = <0xf8014000 0x100>;
-				interrupts = <10 4 6>;
-				dma = <&dma1 0x10002256>;
-				#address-cells = <1>;
-				#size-cells = <0>;
-				pinctrl-names = "default";
-				pinctrl-0 = <&pinctrl_i2c1>;
-				status = "disabled";
-			};
-
-			i2c2: i2c@f8018000 {
-				compatible = "atmel,at91sam9x5-i2c";
-				reg = <0xf8018000 0x100>;
-				interrupts = <11 4 6>;
-				dma = <&dma0 0x1000229A>;
-				#address-cells = <1>;
-				#size-cells = <0>;
-				pinctrl-names = "default";
-				pinctrl-0 = <&pinctrl_i2c2>;
-				status = "disabled";
-			};
-
-			adc0: adc@f804c000 {
-				compatible = "atmel,at91sam9260-adc";
-				reg = <0xf804c000 0x100>;
-				interrupts = <19 4 0>;
-				atmel,adc-use-external;
-				atmel,adc-channels-used = <0xffff>;
-				atmel,adc-vref = <3300>;
-				atmel,adc-num-channels = <12>;
-				atmel,adc-startup-time = <40>;
-				atmel,adc-channel-base = <0x50>;
-				atmel,adc-drdy-mask = <0x1000000>;
-				atmel,adc-status-register = <0x30>;
-				atmel,adc-trigger-register = <0xc0>;
-
-				trigger@0 {
-					trigger-name = "external-rising";
-					trigger-value = <0x1>;
-					trigger-external;
-				};
-
-				trigger@1 {
-					trigger-name = "external-falling";
-					trigger-value = <0x2>;
-					trigger-external;
-				};
-
-				trigger@2 {
-					trigger-name = "external-any";
-					trigger-value = <0x3>;
-					trigger-external;
-				};
-
-				trigger@3 {
-					trigger-name = "continuous";
-					trigger-value = <0x6>;
-				};
-			};
-
-			mmc0: mmc@f0008000 {
-				compatible = "atmel,hsmci";
-				reg = <0xf0008000 0x600>;
-				interrupts = <12 4 0>;
-				status = "disabled";
-				#address-cells = <1>;
-				#size-cells = <0>;
-				dma = <&dma0 0x10002200>;
-				pinctrl-names = "default";
-			};
-
-			mmc1: mmc@f000c000 {
-				compatible = "atmel,hsmci";
-				reg = <0xf000c000 0x600>;
-				interrupts = <26 4 0>;
-				status = "disabled";
-				#address-cells = <1>;
-				#size-cells = <0>;
-				dma = <&dma1 0x10002200>;
-				pinctrl-names = "default";
+			pmc: pmc@fffffc00 {
+				compatible = "atmel,at91rm9200-pmc";
+				reg = <0xfffffc00 0x100>;
 			};
 
-			lcd_bus@f8038000 {
-				#address-cells = <1>;
-				#size-cells = <1>;
-				compatible = "atmel,at91sam9x5-lcd-bus", "simple-bus";
-				ranges = <0xf8038000 0xf8038000 0x4000>;
-				pinctrl-names = "default";
-				pinctrl-0 = <&pinctrl_lcd>;
-				status = "disabled";
-
-				lcd@f8038000 {
-					compatible = "atmel,at91sam9x5-lcd";
-					reg = <0xf8038000 0xff
-					       0xf8038400 0x3ff>;
-					interrupts = <25 4 3>;
-					status = "disabled";
-				};
-
-				lcdovl1@f8038100 {
-					compatible = "atmel,at91sam9x5-lcd";
-					reg = <0xf8038100 0xff
-					       0xf8038800 0x3ff>;
-					status = "disabled";
-				};
-
-				lcdheo1@f8038280 {
-					compatible = "atmel,at91sam9x5-heo";
-					reg = <0xf8038280 0xbf>;
-					interrupts = <25 4 3>;
-					status = "disabled";
-				};
+			rstc@fffffe00 {
+				compatible = "atmel,at91sam9g45-rstc";
+				reg = <0xfffffe00 0x10>;
 			};
 
-			spi0: spi@f0000000 {
-				#address-cells = <1>;
-				#size-cells = <0>;
-				compatible = "atmel,at91rm9200-spi";
-				reg = <0xf0000000 0x100>;
-				interrupts = <13 4 3>;
-				cs-gpios = <&pioA 14 0
-					    &pioA 7 0 /* conflicts with TXD2 */
-					    &pioA 1 0 /* conflicts with RXD0 */
-					    &pioB 3 0 /* conflicts with ERXDV */
-					   >;
-				dma-mask = <0xffffffff>;
-				dma_type = <2>;
-				version = <2>;
-				status = "disabled";
+			shdwc@fffffe10 {
+				compatible = "atmel,at91sam9x5-shdwc";
+				reg = <0xfffffe10 0x10>;
 			};
 
-			spi1: spi@f0004000 {
-				#address-cells = <1>;
-				#size-cells = <0>;
-				compatible = "atmel,at91rm9200-spi";
-				reg = <0xf0004000 0x100>;
-				interrupts = <14 4 3>;
-				cs-gpios = <&pioA 8 0 /* conflitcs with RXD2 */
-					    &pioA 0 0 /* conflitcs with TXD0 */
-					    &pioA 31 0 /* conflitcs with TWCK0 */
-					    &pioA 30 0 /* conflitcs with TWD0 */
-					   >;
-				dma-mask = <0xffffffff>;
-				dma_type = <2>;
-				version = <2>;
-				status = "disabled";
+			pit: timer@fffffe30 {
+				compatible = "atmel,at91sam9260-pit";
+				reg = <0xfffffe30 0xf>;
+				interrupts = <1 4 7>;
 			};
 
 			watchdog@fffffe40 {
@@ -635,43 +826,4 @@
 			status = "disabled";
 		};
 	};
-
-	i2c@0 {
-		compatible = "i2c-gpio";
-		gpios = <&pioA 30 0 /* sda */
-			 &pioA 31 0 /* scl */
-			>;
-		i2c-gpio,sda-open-drain;
-		i2c-gpio,scl-open-drain;
-		i2c-gpio,delay-us = <2>;	/* ~100 kHz */
-		#address-cells = <1>;
-		#size-cells = <0>;
-		status = "disabled";
-	};
-
-	i2c@1 {
-		compatible = "i2c-gpio";
-		gpios = <&pioC 0 0 /* sda */
-			 &pioC 1 0 /* scl */
-			>;
-		i2c-gpio,sda-open-drain;
-		i2c-gpio,scl-open-drain;
-		i2c-gpio,delay-us = <2>;	/* ~100 kHz */
-		#address-cells = <1>;
-		#size-cells = <0>;
-		status = "disabled";
-	};
-
-	i2c@2 {
-		compatible = "i2c-gpio";
-		gpios = <&pioB 4 0 /* sda */
-			 &pioB 5 0 /* scl */
-			>;
-		i2c-gpio,sda-open-drain;
-		i2c-gpio,scl-open-drain;
-		i2c-gpio,delay-us = <2>;	/* ~100 kHz */
-		#address-cells = <1>;
-		#size-cells = <0>;
-		status = "disabled";
-	};
 };
diff --git a/arch/arm/boot/dts/at91sam9x5cm.dtsi b/arch/arm/boot/dts/at91sam9x5cm.dtsi
index ca049fb..fa4a12c 100644
--- a/arch/arm/boot/dts/at91sam9x5cm.dtsi
+++ b/arch/arm/boot/dts/at91sam9x5cm.dtsi
@@ -24,6 +24,21 @@
 	};
 
 	ahb {
+		apb {
+			spi0: spi@f0000000 {
+				cs-gpios = <&pioA 14 0>;
+				/*
+				 * Only valid for Embest and Ronetix modules,
+				 * Cogent modules use an at45 dataflash.
+				 */
+				m25p80@0 {
+					compatible = "atmel,at25df321a";
+					spi-max-frequency = <50000000>;
+					reg = <0>;
+				};
+			};
+		};
+
 		nand0: nand@40000000 {
 			nand-bus-width = <8>;
 			nand-ecc-mode = "hw";
diff --git a/arch/arm/boot/dts/at91sam9x5ek.dtsi b/arch/arm/boot/dts/at91sam9x5ek.dtsi
index 0406e0d..620ec3e 100644
--- a/arch/arm/boot/dts/at91sam9x5ek.dtsi
+++ b/arch/arm/boot/dts/at91sam9x5ek.dtsi
@@ -18,6 +18,26 @@
 
 	ahb {
 		apb {
+			adc0: adc@f804c000 {
+				pinctrl-0 = <
+					&pinctrl_adc0_adtrg
+					&pinctrl_adc0_ad4
+					&pinctrl_adc0_ad5
+					&pinctrl_adc0_ad6
+					&pinctrl_adc0_ad9
+				>;
+				atmel,adc-channels-used = <0x0270>;
+				atmel,adc-vref = <3300>;
+			};
+
+			i2c0: i2c@f8010000 {
+				24c512@51 {
+					compatible = "24c512";
+					reg = <0x51>;
+					pagesize = <128>;
+				};
+			};
+
 			pinctrl@fffff400 {
 				board {
 					pinctrl_mmc0_cd: mmc0_cd {
@@ -61,7 +81,6 @@
 
 			macb0: ethernet@f802c000 {
 				phy-mode = "rmii";
-				status = "okay";
 			};
 
 			macb1: ethernet@f8030000 {
diff --git a/arch/arm/configs/at91_dt_defconfig b/arch/arm/configs/at91_dt_defconfig
index b175577..a1556ea 100644
--- a/arch/arm/configs/at91_dt_defconfig
+++ b/arch/arm/configs/at91_dt_defconfig
@@ -50,6 +50,8 @@ CONFIG_IPV6=y
 # CONFIG_INET6_XFRM_MODE_TUNNEL is not set
 # CONFIG_INET6_XFRM_MODE_BEET is not set
 CONFIG_IPV6_SIT_6RD=y
+CONFIG_CAN=y
+CONFIG_CAN_AT91=y
 # CONFIG_WIRELESS is not set
 CONFIG_UEVENT_HELPER_PATH="/sbin/hotplug"
 CONFIG_DEVTMPFS=y
@@ -60,6 +62,9 @@ CONFIG_MTD=y
 CONFIG_MTD_CMDLINE_PARTS=y
 CONFIG_MTD_CHAR=y
 CONFIG_MTD_BLOCK=y
+CONFIG_MTD_CFI=y
+CONFIG_MTD_DATAFLASH=y
+CONFIG_MTD_M25P80=y
 CONFIG_MTD_NAND=y
 CONFIG_MTD_NAND_ATMEL=y
 CONFIG_MTD_UBI=y
@@ -71,7 +76,8 @@ CONFIG_BLK_DEV_RAM_COUNT=4
 CONFIG_BLK_DEV_RAM_SIZE=8192
 CONFIG_ATMEL_PWM=y
 CONFIG_ATMEL_TCLIB=y
-CONFIG_EEPROM_93CX6=m
+CONFIG_EEPROM_AT24=y
+CONFIG_EEPROM_AT25=y
 CONFIG_SCSI=y
 CONFIG_BLK_DEV_SD=y
 CONFIG_SCSI_MULTI_LUN=y
@@ -99,19 +105,22 @@ CONFIG_INPUT_MOUSEDEV_SCREEN_Y=272
 CONFIG_INPUT_JOYDEV=y
 CONFIG_INPUT_EVDEV=y
 # CONFIG_KEYBOARD_ATKBD is not set
+CONFIG_KEYBOARD_QT1070=y
 CONFIG_KEYBOARD_GPIO=y
 # CONFIG_INPUT_MOUSE is not set
 CONFIG_INPUT_TOUCHSCREEN=y
+CONFIG_TOUCHSCREEN_ATMEL_TSADCC=y
 # CONFIG_SERIO is not set
 CONFIG_LEGACY_PTY_COUNT=4
 CONFIG_SERIAL_ATMEL=y
 CONFIG_SERIAL_ATMEL_CONSOLE=y
 CONFIG_HW_RANDOM=y
 CONFIG_I2C=y
+CONFIG_I2C_CHARDEV=y
+CONFIG_I2C_AT91=y
 CONFIG_I2C_GPIO=y
 CONFIG_SPI=y
 CONFIG_SPI_ATMEL=y
-CONFIG_PINCTRL_AT91=y
 # CONFIG_HWMON is not set
 CONFIG_WATCHDOG=y
 CONFIG_AT91SAM9X_WATCHDOG=y
@@ -119,10 +128,10 @@ CONFIG_SSB=m
 CONFIG_FB=y
 CONFIG_FB_MODE_HELPERS=y
 CONFIG_FB_ATMEL=y
+CONFIG_FB_ATMEL_HLCD=y
 CONFIG_BACKLIGHT_LCD_SUPPORT=y
 # CONFIG_LCD_CLASS_DEVICE is not set
 CONFIG_BACKLIGHT_CLASS_DEVICE=y
-CONFIG_BACKLIGHT_ATMEL_LCDC=y
 # CONFIG_BACKLIGHT_GENERIC is not set
 CONFIG_FRAMEBUFFER_CONSOLE=y
 CONFIG_FRAMEBUFFER_CONSOLE_DETECT_PRIMARY=y
@@ -131,11 +140,16 @@ CONFIG_FONT_8x8=y
 CONFIG_FONT_ACORN_8x8=y
 CONFIG_FONT_MINI_4x6=y
 CONFIG_LOGO=y
-# CONFIG_HID_SUPPORT is not set
+CONFIG_SOUND=y
+CONFIG_SND=y
+# CONFIG_SND_DRIVERS is not set
+CONFIG_SND_ATMEL_AC97C=y
+# CONFIG_SND_SPI is not set
+# CONFIG_SND_USB is not set
+CONFIG_SND_SOC=y
+CONFIG_SND_ATMEL_SOC=y
 CONFIG_USB=y
 CONFIG_USB_ANNOUNCE_NEW_DEVICES=y
-CONFIG_USB_DEVICEFS=y
-# CONFIG_USB_DEVICE_CLASS is not set
 CONFIG_USB_EHCI_HCD=y
 CONFIG_USB_OHCI_HCD=y
 CONFIG_USB_ACM=y
@@ -166,13 +180,17 @@ CONFIG_RTC_CLASS=y
 CONFIG_RTC_DRV_AT91RM9200=y
 CONFIG_RTC_DRV_AT91SAM9=y
 CONFIG_DMADEVICES=y
+CONFIG_AT_HDMAC=y
 # CONFIG_IOMMU_SUPPORT is not set
+CONFIG_IIO=y
+CONFIG_AT91_ADC=y
 CONFIG_EXT2_FS=y
 CONFIG_FANOTIFY=y
 CONFIG_VFAT_FS=y
 CONFIG_TMPFS=y
+CONFIG_JFFS2_FS=y
+CONFIG_UBIFS_FS=y
 CONFIG_NFS_FS=y
-CONFIG_NFS_V3=y
 CONFIG_ROOT_NFS=y
 CONFIG_NLS_CODEPAGE_437=y
 CONFIG_NLS_CODEPAGE_850=y
@@ -183,7 +201,6 @@ CONFIG_DEBUG_FS=y
 # CONFIG_DEBUG_BUGVERBOSE is not set
 # CONFIG_FTRACE is not set
 CONFIG_DEBUG_USER=y
-CONFIG_CRYPTO=y
 CONFIG_CRYPTO_ECB=y
 CONFIG_CRYPTO_AES=y
 CONFIG_CRYPTO_ARC4=y
diff --git a/arch/arm/mach-at91/at91sam9x5.c b/arch/arm/mach-at91/at91sam9x5.c
index a963306..fc4b02b 100644
--- a/arch/arm/mach-at91/at91sam9x5.c
+++ b/arch/arm/mach-at91/at91sam9x5.c
@@ -222,30 +222,35 @@ static struct clk *periph_clocks[] __initdata = {
 
 static struct clk_lookup periph_clocks_lookups[] = {
 	/* lookup table for DT entries */
-	CLKDEV_CON_DEV_ID("usart", "fffff200.serial", &mck),
+	CLKDEV_CON_DEV_ID("hclk", "600000.ohci", &uhphs_clk),
+	CLKDEV_CON_DEV_ID("ohci_clk", "600000.ohci", &uhphs_clk),
+	CLKDEV_CON_DEV_ID("ehci_clk", "700000.ehci", &uhphs_clk),
+	CLKDEV_CON_DEV_ID("spi_clk", "f0000000.spi", &spi0_clk),
+	CLKDEV_CON_DEV_ID("spi_clk", "f0004000.spi", &spi1_clk),
+	CLKDEV_CON_DEV_ID("mci_clk", "f0008000.mmc", &mmc0_clk),
+	CLKDEV_CON_DEV_ID("mci_clk", "f000c000.mmc", &mmc1_clk),
+	CLKDEV_CON_DEV_ID("can_clk", "f8000000.can", &can0_clk),
+	CLKDEV_CON_DEV_ID("can_clk", "f8004000.can", &can1_clk),
+	CLKDEV_CON_DEV_ID("t0_clk", "f8008000.timer", &tcb0_clk),
+	CLKDEV_CON_DEV_ID("t0_clk", "f800c000.timer", &tcb0_clk),
+	CLKDEV_CON_DEV_ID(NULL, "f8010000.i2c", &twi0_clk),
+	CLKDEV_CON_DEV_ID(NULL, "f8014000.i2c", &twi1_clk),
+	CLKDEV_CON_DEV_ID(NULL, "f8018000.i2c", &twi2_clk),
 	CLKDEV_CON_DEV_ID("usart", "f801c000.serial", &usart0_clk),
 	CLKDEV_CON_DEV_ID("usart", "f8020000.serial", &usart1_clk),
 	CLKDEV_CON_DEV_ID("usart", "f8024000.serial", &usart2_clk),
 	CLKDEV_CON_DEV_ID("usart", "f8028000.serial", &usart3_clk),
-	CLKDEV_CON_DEV_ID("t0_clk", "f8008000.timer", &tcb0_clk),
-	CLKDEV_CON_DEV_ID("t0_clk", "f800c000.timer", &tcb0_clk),
+	/* additional fake clock for macb_hclk */
+	CLKDEV_CON_DEV_ID("hclk", "f802c000.ethernet", &macb0_clk),
+	CLKDEV_CON_DEV_ID("hclk", "f8030000.ethernet", &macb1_clk),
+	CLKDEV_CON_DEV_ID("tsc_clk", "f804c000.tsadcc", &adc_clk),
 	CLKDEV_CON_DEV_ID("dma_clk", "ffffec00.dma-controller", &dma0_clk),
 	CLKDEV_CON_DEV_ID("dma_clk", "ffffee00.dma-controller", &dma1_clk),
-	CLKDEV_CON_DEV_ID("mci_clk", "f0008000.mmc", &mmc0_clk),
-	CLKDEV_CON_DEV_ID("mci_clk", "f000c000.mmc", &mmc1_clk),
-	CLKDEV_CON_DEV_ID(NULL, "f8010000.i2c", &twi0_clk),
-	CLKDEV_CON_DEV_ID(NULL, "f8014000.i2c", &twi1_clk),
-	CLKDEV_CON_DEV_ID(NULL, "f8018000.i2c", &twi2_clk),
+	CLKDEV_CON_DEV_ID("usart", "fffff200.serial", &mck),
 	CLKDEV_CON_DEV_ID(NULL, "fffff400.gpio", &pioAB_clk),
 	CLKDEV_CON_DEV_ID(NULL, "fffff600.gpio", &pioAB_clk),
 	CLKDEV_CON_DEV_ID(NULL, "fffff800.gpio", &pioCD_clk),
 	CLKDEV_CON_DEV_ID(NULL, "fffffa00.gpio", &pioCD_clk),
-	/* additional fake clock for macb_hclk */
-	CLKDEV_CON_DEV_ID("hclk", "f802c000.ethernet", &macb0_clk),
-	CLKDEV_CON_DEV_ID("hclk", "f8030000.ethernet", &macb1_clk),
-	CLKDEV_CON_DEV_ID("hclk", "600000.ohci", &uhphs_clk),
-	CLKDEV_CON_DEV_ID("ohci_clk", "600000.ohci", &uhphs_clk),
-	CLKDEV_CON_DEV_ID("ehci_clk", "700000.ehci", &uhphs_clk),
 };
 
 /*
diff --git a/drivers/net/can/Kconfig b/drivers/net/can/Kconfig
index 2edfe34..498926e 100644
--- a/drivers/net/can/Kconfig
+++ b/drivers/net/can/Kconfig
@@ -56,7 +56,7 @@ config CAN_CALC_BITTIMING
 
 config CAN_AT91
 	tristate "Atmel AT91 onchip CAN controller"
-	depends on CAN_DEV && (ARCH_AT91SAM9263 || ARCH_AT91SAM9X5 || SOC_SAMA5D3)
+	depends on CAN_DEV && ARCH_AT91
 	---help---
 	  This is a driver for the SoC CAN controller in Atmel's AT91SAM9263,
 	  AT91SAM9X5 and SAMA5D3 processors.
-- 
1.7.10.4

